fix: registration modal fix to ensure complete decoupling

bradymiller opened the pull request:
fixes #8334

~~(prob gonna do a little more work to make it clear what getProductStatus() is for by renaming it to getProductStatusDialog() and then offshooting the call from the about page to a small new function for collecting the email adress)~~ DONE

bradymiller commented:
@sjpadgett ,
I think this will do. I should of made my prior work more clear that basically hijacked that function for the dialog display :)
I've just got to test it a bit more (had to go to sleep :) ).

sjpadgett commented:
cool, I'll test some more also tomorrow. tired here as well, good night.

sjpadgett commented on src/Services/ProductRegistrationService.php:
```diff
@@ -55,21 +56,27 @@ public function getProductStatus(): array
+                if ($optOut == null) {
+                    $row['allowEmail'] = 1;
+                }
+                if ($telemetry_disabled == null) {
+                    $row['allowTelemetry'] = 1;
```
@bradymiller In current actions, $row['allowTelemetry'] should be enabled if $telemetry_disabled == 0 otherwise, none of the event collection routines will run to report telemetry.
Recommend a new flag collectTelemetryEvents or something. It'd be used in three places currently portal JS, Top Nav from tabs_view_model.js and API dispatcher.

sjpadgett replied:
Just a reminder to remove the exposed unused 
```
case 'allUsageData':
            $result = $telemetryService->reportUsageData();
            echo json_encode($result);
            break;
```

In track_portal_events.php

bradymiller replied:
will remove that chunk of code in commit to PR tonight :)

bradymiller replied:
The three places that assess if telemetry is disabled now use your handydandy TelemetryService::isTelemetryEnabled() call :)

sjpadgett replied:
when in doubt, write a method!:)

bradymiller commented on interface/main/tabs/main.php:
```diff
@@ -121,7 +122,7 @@
     const isFax = "<?php echo !empty($GLOBALS['oefax_enable_fax']) ?? null?>";
     const isServicesOther = (isSms || isFax);
-    const allowTelemetry = <?php echo js_escape($allowTelemetry); ?>;
+    var telemetryEnabled = <?php echo js_escape(TelemetryService::isTelemetryEnabled()); ?>;
```
@sjpadgett , for some reason I needed to have this set as a var for it to be seen by tabs_view_model.js script. There are some comments above about how let/const may be out of scope when try to call from other places. This also allows it to be modified (ie. would be nice to turn it on when opt-in rather than needing to logout/login; this is very optional since only gonna gain a little bit of data, but have the option to add this).

sjpadgett replied:
I put that note in back when I did the jsGlobals!

bradymiller commented on src/Telemetry/TelemetryService.php:
```diff
@@ -42,7 +42,7 @@ public static function isTelemetryEnabled(): int
         // Check if telemetry is disabled in the product registration table.
         $isEnabled = sqlQuery("SELECT `telemetry_disabled` FROM `product_registration` WHERE `telemetry_disabled` = 0")['telemetry_disabled'] ?? null;
-        if ((int)$isEnabled === 0) {
+        if (!is_null($isEnabled)) {
```
@sjpadgett , This was returning 1 no matter what. I am guessing that forcing null to into makes it a zero. The above change works nicely on testing (when no entry, when null, when 0, and when 1).

sjpadgett replied:
yeah I should know better. lol I just told Ruth about this!

bradymiller commented:
hi @sjpadgett ,
Couple things I noted on testing:
- need to scrub the web path from the portal event_url in the event record for portal events
- noted that api scopes are stored in event_label api event record. will this break the varchar 255 when lots of scopes?

sjpadgett commented:
see here. is this not enough?

        $eventUrl = preg_replace('/\?.*$/', '', $data['eventUrl'] ?? '');

bradymiller commented:
1. Looks like event_label is varchar 255 in database.sql and upgrade script

2. http://localhost:8300/portal/patient/onsitedocuments showing in my portal event_url (note full path not showing in the api or core menu events)

sjpadgett commented:
> 1. Looks like event_label is varchar 255 in database.sql and upgrade script
> 2. http://localhost:8300/portal/patient/onsitedocuments showing in my portal event_url (note full path not showing in the api or core menu events)

event_label needs to be varchar because it is a unique key. MariaDB or MySql(forget) doesn't like TEXT as unique keys so I changed in master.
http://localhost:8300/portal/patient/onsitedocuments is by design unless PHI is sneaking by. Some of the portal events I want the drill down and the card structures were engineered to be easy to id cards thus the *-go.

sjpadgett commented:
Scopes are saved to eventTarget a text column.

```

// Report to Telemetry
// Be nice to find a centralized place to catch failed requests.
try {
    (new TelemetryService())->trackApiRequestEvent([
        'eventType' => 'API',
        'eventLabel' => strtoupper($_SESSION['api'] ?? 'UNKNOWN'),
        'eventUrl' => $restRequest->getRequestMethod() . ' ' . $resource,
        'eventTarget' => json_encode($GLOBALS['oauth_scopes'] ?? []),
    ]);
    exit;
```

sjpadgett commented:
@bradymiller Nope not weird simply I missed a commit, sorry. I don't think I missed any others but I'll check. Whatever you think is best for portal url maybe normalize the domain!
```

// Report to Telemetry
// Be nice to find a centralized place to catch failed requests.
try {
    (new TelemetryService())->trackApiRequestEvent([
        'eventType' => 'API',
        'eventLabel' => strtoupper($_SESSION['api'] ?? 'UNKNOWN'),
        'eventUrl' => $restRequest->getRequestMethod() . ' ' . $resource,
        'eventTarget' => json_encode($GLOBALS['oauth_scopes'] ?? []),
    ]);
    exit;
} catch (\Exception $e) {
    $logger->error("dispatch.php telemetry error", ['exception' => $e]);
}

```

bradymiller commented on apis/dispatch.php:
```diff
@@ -414,9 +414,9 @@
-        'eventTarget' => $restRequest->getRequestMethod() . ' ' . $_SESSION['api'],
+        'eventLabel' => strtoupper($_SESSION['api'] ?? 'UNKNOWN'),
+        'eventUrl' => $restRequest->getRequestMethod() . ' ' . $resource,
+        'eventTarget' => json_encode($GLOBALS['oauth_scopes'] ?? []),
```
@sjpadgett , just put in your code for this fix. Also made the parameter type an array instead of mixed. and dropped reportUsageData case from track_events.php since realized not being used anywhere.

bradymiller commented:
btw, above testing well :)

sjpadgett commented:
@bradymiller Use if you want

TelemetryService.php L-73
// Sanitize URL by stripping query parameters and location parts.
$eventUrl = $this->normalizeUrl($data['eventUrl'] ?? '');

```

    function normalizeUrl(string $url): string {
        // query string is not needed for telemetry
        $parsed = parse_url($url);
        $path = $parsed['path'] ?? '';
        $fragment = isset($parsed['fragment']) ? '#' . $parsed['fragment'] : '';
        $normalized = preg_replace('#^/(interface|portal|modules|sites|openemr)?#', '', $path);

        return ltrim($normalized . $fragment, '/');
    }
```

bradymiller commented on src/Telemetry/TelemetryService.php:
```diff
@@ -179,10 +183,23 @@ public function reportUsageData(): int|bool
+            $normalized = $path;
+        }
+        return ($normalized . $fragment);
+    }
```
@sjpadgett , this is working nicely and basically look like urls that core is bringing in now. only doing this on the portal urls. 
Thank goodness had mechanism to test webroot via easy dev:
```
webroot-management:
  cwb, change-webroot-blank          Change webroot to be blank (this is default setting of environment)
  cwo, change-webroot-openemr        Change webroot to be openemr
```

sjpadgett replied:
yep empty webroot is an issue that bites us all the time for those that alias elsewhere. I try to remember to test on localhost but still get caught. Nice catch!

bradymiller commented:
Added debug log support for events and couldn't help myself and added support to turn on the top.telemetryEnabled flag when consent for telemetry (so do not need to logout/login for it to start collecting telemetry).

sjpadgett commented:
We may want to collect what API grant type is being used.

bradymiller commented:
Testing well and bringing this in. Gonna need a break from this for at least 24 hours (testing all the different possibilities is a bit crazy :) ). Will also stage this on the PR for patch 4.
regarding the api grant stuff, another day, another PR :)

bradymiller commented:
Brought in rel-703 for patch 7.0.3.4: #8338
